### ISSUE TITLE
Support setting guestInfo properties on extraConfig

### ISSUE THREAD

**Issue by rdavisunr** [vmware/terraform-provider-vcd#822]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I am using Terraform to provision Windows VMs in both vSphere and VCD environments. Due to the Windows requirement, I am using [cloudbase-init](https://cloudbase-init.readthedocs.io/en/latest/intro.html) for instance initialization.

In cloudbase-init, I am using the VMWare GuestInfo [datasource](https://cloudbase-init.readthedocs.io/en/latest/services.html#vmware-guestinfo-service) (which has similar support in [cloud-init](https://cloudinit.readthedocs.io/en/latest/topics/datasources/vmware.html)). These datasources read values from "guestInfo" keys, set on the **extraConfig** property (reference this [Data Object description](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/723e7f8b-4f21-448b-a830-5f22fd931b01/5a8257bd-7f41-4423-9a73-03307535bd42/doc/vim.vm.ConfigInfo.html)).

When using the [vSphere provider](https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs), this all works flawlessly. The [vsphere_virtual_machine](https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine) resource provides a means to directly set values on [extra_config](https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#extra_config). The cloudinit [docs ](https://cloudinit.readthedocs.io/en/latest/topics/datasources/vmware.html)outline the exact guestinfo keys to be set on extraConfig.

However, no such mechanism is available for the [vcd_vapp_vm](https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/vapp_vm) resource. As a consequence, I tried using the [guest_properties](https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/vapp_vm#guest_properties) arguments, but this is incompatible with the [VMWare GuestInfo datasource](https://cloudbase-init.readthedocs.io/en/latest/services.html#vmware-guestinfo-service) since it doesn't seem to set properties on extraConfig.

Following suggestions from other issues in this repo (like [this one](https://github.com/vmware/terraform-provider-vcd/issues/510#issuecomment-843721455)), I tried switching to DataSourceOvf. However, cloudbase-init doesn't support reading the vApp properties exposed over the ovf.env variable. It seems that cloudbase-init can read OVF from a mounted ISO, but this doesn't help me as I need to pass the values via Terraform.

Since I have to use cloudbase-init, I may also log an issue in their repo to see if they could expand their OVF support to include vApp properties.  I also am outlining a workaround below.

However, since both cloudbase-init and cloud-init have a VMWare datasource for extraConfig/guestInfo, it would be very beneficial for users if the VCD provider could support setting properties on extraConfig as the vsphere provider does.

### New or Affected Resource(s)

* vcd_vapp_vm
* vcd_vm

### Terraform Configuration

Just for reference, this snippet shows how I am setting extraConfig properties in the vSphere provider.

```hcl
extra_config = {
  "guestinfo.metadata" = base64gzip(jsonencode({
    "instance-id": local.vm_name,
    "local-hostname": var.computer_name,
    "admin-username": "Administrator",
    "admin-password": var.admin_password,
  })),
  "guestinfo.metadata.encoding" = "gzip+base64",
  "guestinfo.userdata" = data.template_cloudinit_config.cloudinit.rendered,
  "guestinfo.userdata.encoding" = "gzip+base64"
}
```

### Workaround
For now, I am using a local-exec provisioner that calls a Powershell script to set the appropriate values. My script is similar to the script show here: [GuestInfo.ps1](https://github.com/lamw/vmware-scripts/blob/master/powershell/GuestInfo.ps1)

However, this is problematic in a number of different ways. For example, when creating/modifying the VM, I have to have the [power_on](https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/vapp_vm#power_on) state set to false, so the local-exec provisioner can set the properties on extraConfig before the VM boots. Then, the VM has to be subsequently powered on.

### References
* [vSphere provider extraConfig ](https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#extra_config)
* [Data Object](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/723e7f8b-4f21-448b-a830-5f22fd931b01/5a8257bd-7f41-4423-9a73-03307535bd42/doc/vim.vm.ConfigInfo.html) that includes extraConfig
* cloudbase-init VMWare data source Service [docs ](https://cloudbase-init.readthedocs.io/en/latest/services.html#vmware-guestinfo-service)and [source code](https://github.com/cloudbase/cloudbase-init/blob/f8479e5cbc27c88f4fcf9a781d7cd5d6b7a9ff74/cloudbaseinit/metadata/services/vmwareguestinfoservice.py)
* clousebase-init OVF  data source Service [docs](https://cloudbase-init.readthedocs.io/en/latest/services.html#open-virtualization-format-ovf) and [source code](https://github.com/cloudbase/cloudbase-init/blob/f8479e5cbc27c88f4fcf9a781d7cd5d6b7a9ff74/cloudbaseinit/metadata/services/ovfservice.py) (which only supports ISO transport)
* cloud-init VMWare data source [docs](https://cloudinit.readthedocs.io/en/latest/topics/datasources/vmware.html) and [source code](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceVMware.py)
* cloud-init OVF data source [docs](https://cloudinit.readthedocs.io/en/latest/topics/datasources/ovf.html) and [source code](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceVMware.py) (which supports both ISO and VMware tools transports)

**Comment by Didainius** [comment]
Hello @rdavisunr ,
I think you are looking what is called in VCD Guest properties. https://registry.terraform.io/providers/vmware/vcd/latest/docs/data-sources/vapp_vm#guest_properties

That way you could feed it in. Have a look at our docs, which are no exactly your case, but still use guest properties: https://registry.terraform.io/providers/vmware/vcd/latest/docs/guides/vm_guest_customization

**Comment by Didainius** [comment]
I see you have tried to use it. But it does look to be the same to me. Could you check if you can achieve what you want using VCD UI?

**Comment by Didainius** [comment]
Here is also a snippet of how one can configure Ubuntu using cloud-init. It just uses different datasource, but still this works for ubuntu 21.10+ IIRC:
```
guest_properties = {
  "instance-id" = format("%s-terminal", var.username)
  "hostname"    = format("%s-terminal", var.username)
  "public-keys" = var.ssh-public-key
  "user-data"   = base64encode(var.setup-script)
}

```

**Comment by rdavisunr** [comment]
@Didainius - thanks for the response!

vcd_vapp_vm seems to be [setting](https://github.com/vmware/terraform-provider-vcd/blob/89bcdc8abb7a31d33e972b02f94a077da93ccb93/vcd/resource_vcd_vapp_vm.go#L2258) the guest properties on the [ProductSection elements](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/6ac8164c-8844-4188-ac1b-cd59721c06b8/7d36e490-310e-4485-91cc-c3abb23b0d32/GUID-E13A5613-8A41-46E3-889B-8E1EAF10ABBE.html).

That approach is fine for datasources that can read the OVF environment. Unfortunately, cloudbase-init doesn't have a datasource that can read the ovf env. Their only available OVF datasource can read from an ISO only, which won't work for me :(

So, I must use their vmware guestinfo datasource. But, the cloudbase-init [VMWare data source](https://cloudbase-init.readthedocs.io/en/latest/services.html#vmware-guestinfo-service) isn't reading from the ProductSection guest properties elements. Rather, it (and the similar cloud-init [datasource](https://cloudinit.readthedocs.io/en/latest/topics/datasources/vmware.html)) read these values off the extraConfig property.

This extraConfig property (look for it on [this page](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/723e7f8b-4f21-448b-a830-5f22fd931b01/5a8257bd-7f41-4423-9a73-03307535bd42/doc/vim.vm.ConfigInfo.html)) seems to be different from the ProductSection elements.

I believe that's why the vSphere provider argument is named extraConfig.

So, the question becomes...

Is it possible for the VCD provider to set values on the extraConfig property on the vcd_vapp_vm resource?

I found a govc example [here ](https://github.com/vmware/govmomi/blob/master/govc/USAGE.md#vmchange)- please note the -e argument, which allows properties to be set on extraConfig.

**Comment by Didainius** [comment]
@rdavisunr ,
Did you see that functionality in VCD UI anywhere?

**Comment by rdavisunr** [comment]
@Didainius 

As far as I can tell, no. Unless I am missing some permissions that make it possible.

However, it's possible to do it with the vSphere UI, 

1. Search for the vcloud VM
2. Right click --> Edit Settings
3. Select VM Options Tab
4. Click the Advanced Section
5. In Configuration Parameters, click the Edit Configuration button.

Here's a screenshot:
![image](https://user-images.githubusercontent.com/8481440/163063274-9601a63e-85c5-4026-abb0-0a159fb62628.png)

I am guessing this doesn't bode well for it being possible to accomplish with the go-cloud-director api...

Thanks,

**Comment by vbauzysvmware** [comment]
Hi @[rdavisunr]

Seems there is a way to access extra config through VCD API. So we can add this feature in the backlog.

**Comment by rdavisunr** [comment]
@vbauzysvmware and @Didainius - thank you; that's great news. I will look forward to seeing this in a future release!

**Comment by Blankf** [comment]
Same here, looking forward to this fixes!
would love to be able to deploy windows vms with cloudbase-init

**Comment by rdavisunr** [comment]
@Blankf - here's the workaround repo: https://github.com/rdavisunr/set-vm-guestinfo

Thanks

**Comment by mnspodrska** [comment]
As much as I support that Terraform can configure extraConfig, I'm wondering if there is a way to change OVF environment transport to ISO instead of VMware Tools so that cloudbase-init can use Guest Properties configured in VCD.

If I manually change that setting in vSphere to:
![image](https://user-images.githubusercontent.com/17175348/170537912-87a853fd-9043-4e0d-81c7-368d7bd2d5aa.png)
Cloudbase can find OVF ENV cdrom containing ovf-env.xml with Guest Properties configured in VCD and everything runs smoothly with DataSourceOvf.

@vbauzysvmware can you please check if VCD API has that option so that we can use terraform-provider-vcd to configure it?

**Comment by vbauzysvmware** [comment]
Hi @mnspodrska 

seems not possible, but I still checking some clarifications.

**Comment by prologic** [comment]
I _may_ be barking up the wrong tree here, but I'm struggling myself to get Cloudinit to operate correctly on an Ubuntu 20.04 Guest in VCD. I also set guest properties very similarly to comments above:

```#!hcl
    guest_properties = {
      "hostname"  = "test"
      "meta-data" = base64encode(templatefile("metadata.tmpl", {
      }))
      "user-data" = base64encode(templatefile("userdata.tmpl", {
        "hostname"        = "test",
        "ssh_keys"        = join("\n", var.ssh_keys),
        "loki_logging"    = var.loki_logging,
        "default_ssh_key" = var.default_ssh_key,
      }))
    }
```

But upon creating the resources and booting the vApp/VM Cloudinit doesn't find any valid dadta sources.

Near as I can tell `vmware-rpctool` returns "No value" for what I'm expecting:

```#!sh
root@packer-control-ubuntu-2004-001:~# vmware-rpctool 'info-get guestinfo.userdata'
No value found
root@packer-control-ubuntu-2004-001:~# vmware-rpctool 'info-get guestinfo.user-data'
No value found
```

**Comment by Didainius** [comment]
Hello @prologic ,
From my tests I know that Ubuntu 22.04 had issues with the customization and I surelly know that it started working in 21.10 and 22.04LTS. 

In fact we even improved our Guides page docuimentation with Ubuntu example, which is going to be released as part of 3.7.0. You can have a look at this guides doc which is verified and working here https://github.com/vmware/terraform-provider-vcd/blob/main/website/docs/guides/vm_guest_customization.html.markdown#example-using-cloudinit-configuration-ubuntu-2204-lts

**Comment by prologic** [comment]
Thanks! I'll have a look in the morning 👌

**Comment by jaymzmac** [comment]
Just adding a bit more information to this issue. 

It is possible to use the VCD API to set guestinfo properties on a VM using the [reconfigureVM](https://developer.vmware.com/apis/1703/doc/operations/POST-ReconfigureVm.html) action. 

As an example, the VCD CSE solution uses this to set the `guestinfo.userdata` and `guestinfo.userdata.encoding` properties on the Kubernetes nodes that it spins up:
https://github.com/vmware/cloud-provider-for-cloud-director/blob/main/pkg/vcdsdk/vapp.go#L494
https://github.com/vmware/cluster-api-provider-cloud-director/blob/main/controllers/vcdmachine_controller.go#L760

I understand that there is a documented _solution_ that leverages `guestinfo.ovfEnv` and the cloud-init OVF datasource, but it would be really good to instead have a way of setting the `guestinfo.userdata`, `guestinfo.userdata.encoding`, `guestinfo.metadata`, `guestinfo.metadata.encoding` properties directly with terraform. 

The VMware GuestInfo datasource is far superior to the OVF datasource in my opinion, with many solutions leveraging it. And as the OP stated, Cloudbase-Init (the Windows equivalent of Cloud-Init) also has a VMware GuestInfo datasource that uses the same guestinfo properties for initializing Windows VMs. So adding this functionality would allow both Linux VMs and Windows VMs to be initialized in the same fashion.

**Comment by forestEagle** [comment]
> As much as I support that Terraform can configure extraConfig, I'm wondering if there is a way to change OVF environment transport to ISO instead of VMware Tools so that cloudbase-init can use Guest Properties configured in VCD.
> 
> If I manually change that setting in vSphere to: ![image](https://user-images.githubusercontent.com/17175348/170537912-87a853fd-9043-4e0d-81c7-368d7bd2d5aa.png) Cloudbase can find OVF ENV cdrom containing ovf-env.xml with Guest Properties configured in VCD and everything runs smoothly with DataSourceOvf.
> 
> @vbauzysvmware can you please check if VCD API has that option so that we can use terraform-provider-vcd to configure it?


Been looking for a way of doing this as well, since open-vm-tools isn't installed by default on the Debian 12 cloudimages, and it seems to be the only way to bootstrap vms with the cloud-init/user-data without modifying this value.
 However, since I already need to convert the qcow2 to a ova,
`qemu-img convert -f qcow2 -O vmdk debian-12-genericcloud-amd64.qcow2 debian-12-genericcloud-amd64_temp.vmdk `
`vmdk-convert debian-12-genericcloud-amd64_temp.vmdk debian-12-genericcloud-amd64.vmdk `
`mkova.sh debian-12-genericcloud-amd64 debian-12-genericcloud-amd64.vmdk -hw 19`

I've modified the ovf template for the relevant hardware version in the [open-vmdk template](https://github.com/vmware/open-vmdk/tree/master/templates) and replaced `<VirtualHardwareSection>` with `<VirtualHardwareSection ovf:transport="iso">`. After this, I upload the template to the content library and the cloned vm's retain this 'vApp setting'.
This is enough to bootstrap the VM with the user-data setting as mentioned, and I even install the VMware tools service to retrieve the DHCP assigned IPv4 in the deployment.

Since OVA's are tarballs it should be possible to edit the OVA after the fact as long as you update the manifest file with the new shasum (haven't really tried though).
You could also use virt-customize to install the package directly to the qcow2 before converting to vmdk/ova but I wanted to avoid this.

**Comment by dataclouder** [comment]
Implemented in PR #1253
